File: algorithms/1_easy/C/01790_check_if_one_string_swap_can_make_strings_equal/solution.c
bool areAlmostEqual(const char *s1, const char *s2) {
    const size_t s1_len = strlen(s1);
    const size_t s2_len = strlen(s2);
    if (s1_len != s2_len) { return false; }

    // cnt short-circuits at >2 mismatches, so each freq is only updated for
    // at most two indices; the deltas stay within [-2, 2]. i8 is safe.
    int8_t freqs[256] = { 0 };
    int cnt = 0;

    for (size_t i = 0; i < s1_len; i += 1) {
        if (s1[i] != s2[i]) {
            cnt += 1;
            if (cnt > 2) {
                return false;
            }
            freqs[(unsigned char)s1[i]] += 1;
            freqs[(unsigned char)s2[i]] -= 1;
        }
    }

    for (size_t i = 0; i < 256; i += 1) {
        if (freqs[i] != 0) {
            return false;
        }
    }

    return true;
}

File: algorithms/1_easy/C/01790_check_if_one_string_swap_can_make_strings_equal/solution.rs
impl Solution {
    pub fn are_almost_equal(s1: String, s2: String) -> bool {
        if s1.len() != s2.len() { return false; }
        let (bytes1, bytes2) = (s1.as_bytes(), s2.as_bytes());

        // cnt short-circuits at >2 mismatches, so each freq is only updated for
        // at most two indices; the deltas stay within [-2, 2]. i8 is safe.
        let mut freqs = [0_i8; 256];
        let mut cnt = 0_i32;

        for i in 0..bytes1.len() {
            if bytes1[i] != bytes2[i] {
                cnt += 1;
                if cnt > 2 {
                    return false;
                }
                freqs[bytes1[i] as usize] += 1;
                freqs[bytes2[i] as usize] -= 1;
            }
        }

        freqs.into_iter().all(|freq| freq == 0)
    }
}
